### ISSUE TITLE
The image built by paketo with kernel profile does not work

### ISSUE THREAD

**Issue by gkwan-ibm** [paketo-buildpacks/liberty#183]
the server.xml contains cdi-3.0, jsonb-2.0, jsonp-2.0, restfulWS-3.0 features

built the image by the following command
```
pack build --path ./system ^
--env BP_JAVA_APP_SERVER=liberty ^
--env BP_LIBERTY_FEATURES="cdi-3.0 jsonb-2.0 jsonp-2.0 restfulWS-3.0" ^
--env BP_MAVEN_BUILT_ARTIFACT="target/*.[ejw]ar src/main/liberty/config/*" ^
--env BP_LIBERTY_PROFILE=kernel ^
--buildpack paketo-buildpacks/eclipse-openj9 ^
--buildpack paketo-buildpacks/java system:1.0-SNAPSHOT
```

ran the image and got following error
```
CWWKF0033E: The singleton features io.openliberty.servlet.api-5.0 and io.openliberty.servlet.api-3.0 cannot be loaded at the same time.  The configured features restfulWS-3.0 and restfulWS-3.0 include one or more features that cause the conflict. Your configuration is not supported; update server.xml to remove incompatible features.
```

**Comment by gkwan-ibm** [comment]
to reproduce the problem:
- `git clone https://github.com/uhryniuk/draft-guide-containerize-paketo.git`
- `cd draft-guide-containerize-paketo/finish`
- `mvn package`
- `pack config default-builder gcr.io/paketo-buildpacks/builder:base`
- build with kernel profile
```
pack build --path ./system \
--env BP_JAVA_APP_SERVER=liberty \
--env BP_MAVEN_BUILT_ARTIFACT="target/*.[ejw]ar src/main/liberty/config/*" \
--env BP_LIBERTY_PROFILE=kernel \
--buildpack paketo-buildpacks/eclipse-openj9 \
--buildpack paketo-buildpacks/java system:1.0-SNAPSHOT
```
- `docker images` and see
  - `system  1.0-SNAPSHOT  b30a5f24c11e   42 years ago    260MB`
  - It is strange the CREATED shows "42 years ago".
- `docker run --rm -d --name system -p 9080:9080 system:1.0-SNAPSHOT`
- `curl http://localhost:9080/system/properties` and got
    - `curl: (52) Empty reply from server`
- `docker logs system` and got
```
{"type":"liberty_message","host":"607eb4a4b582","ibm_userDir":"\/layers\/paketo-buildpacks_liberty\/open-liberty-runtime-kernel\/usr\/","ibm_serverName":"defaultServer","message":"CWWKF0001E: A feature definition could not be found for cdi-3.0","ibm_threadId":"0000002c","ibm_datetime":"2022-06-29T19:57:01.719+0000","ibm_messageId":"CWWKF0001E","module":"com.ibm.ws.kernel.feature.internal.FeatureManager","loglevel":"ERROR","ibm_sequence":"1656532621719_0000000000008"}
```
- Rebuild by jakartaee9:
```
pack build --path ./system \
--env BP_JAVA_APP_SERVER=liberty \
--env BP_MAVEN_BUILT_ARTIFACT="target/*.[ejw]ar src/main/liberty/config/*" \
--env BP_LIBERTY_PROFILE=jakartaee9 \
--buildpack paketo-buildpacks/eclipse-openj9 \
--buildpack paketo-buildpacks/java system:1.0-SNAPSHOT
```
- `docker run --rm -d --name system -p 9080:9080 system:1.0-SNAPSHOT`
- `curl http://localhost:9080/system/properties` and  java System properties

**Comment by dmikusa-pivotal** [comment]
FYI...

>It is strange the CREATED shows "42 years ago".

This is intentional. It's because image builds are reproducible. By default, it squashes all the timestamps to 0, which ends up being 42 years ago.

If you're using pack 0.26.0+, you can set this flag:

```
      --creation-time string        Desired create time in the output image config. Accepted values are Unix timestamps (e.g., '1641013200'), or 'now'. Platform API version must be at least 0.9 to use this feature
```

which allows you to use a custom time or `now` to have the actual timestamp show up.

**Comment by dmikusa-pivotal** [comment]
OK, so when using the kernel profile, you need to provide the list of features to the buildpack through `BP_LIBERTY_FEATURES`. Then the buildpack will install them. It doesn't look at the `server.xml`. The list from that env variable is passed through to the `bin/featureUtility installFeature` command.

Note, I also had to remove the `BP_MAVEN_BUILT_ARTIFACT`, I'll explain that below.

If you use this command to build, it should work:

```
pack build --path ./system \
--env BP_JAVA_APP_SERVER=liberty \
--env BP_LIBERTY_PROFILE=kernel \
--env BP_LIBERTY_FEATURES='restfulWS-3.0 jsonb-2.0 jsonp-2.0 cdi-3.0 jsp-2.3' \
--buildpack paketo-buildpacks/eclipse-openj9 \
--buildpack paketo-buildpacks/java system:1.0-SNAPSHOT
```

I can then `curl` the server. 

----------------------

If you just want the app to work now, that would be the way to go. Having said that, I think the buildpack can do better.

Right now, we only pull features from the env variable. If there is a user-provided server.xml, that could also contain features and we should be fetching those to install them as well. Not checking server.xml also causes a bug in that if the `BP_LIBERTY_FEATURES` env variable contains the same feature that you have listed in server.xml, the buildpack will add the feature a second time which causes an error when Liberty starts. Rather than fix that, I think we should be just reading the list out of server.xml and merging it with the env variable's list. We can then rewrite the entire `featureManager` block in server.xml with the merged list and we won't have any conflicts.

I'm not super familiar with Liberty, but it also sounds like we might be able to run `featureUtility installServerFeatures <server-name>` and have the tool install the features listed in server.xml. That might be easier, but we'd still have to be careful about adding features to server.xml from `BP_LIBERTY_FEATURES` so that we don't cause duplicates.

I'm going to close this issue as questions are answered, but I'll open a new issue to represent the enhancement work that could be done here.

**Comment by uhryniuk** [comment]
Hi @dmikusa-pivotal
It seems your command is able to build with the `kernel` liberty profile but I ran into other issues.

I tried the command and receive a `200` response when I `curl` the server at `http://localhost:9080`, as expected. I got issues when I tried to `curl` an endpoint on the server at `http://localhost:9080/system/properties`, which  I'd receive a `404` response. 

To be exact this is the response from `curl -I http://localhost:9080/system/properties` is:

```
HTTP/1.1 404 Not Found
X-Powered-By: Servlet/3.1
Content-Type: text/html;charset=ISO-8859-1
$WSEP:
Content-Language: en-US
Connection: Close
Date: Tue, 12 Jul 2022 21:21:46 GMT
```

Here is the output from `docker logs system` when hitting the endpoint:
```
{"type":"liberty_message","host":"a08e0407b034","ibm_userDir":"\/layers\/paketo-buildpacks_liberty\/open-liberty-runtime-kernel\/usr\/","ibm_serverName":"defaultServer","message":"SRVE0190E: File not found: \/system\/properties","ibm_threadId":"00000036","ibm_datetime":"2022-07-12T21:25:32.357+0000","ibm_messageId":"SRVE0190E","module":"com.ibm.ws.webcontainer.extension","loglevel":"WARNING","ibm_methodName":"handleRequest","ibm_className":"com.ibm.ws.webcontainer.extension.DefaultExtensionProcessor","ibm_sequence":"1657661132357_0000000000019","ext_appName":"app","ext_thread":"Default Executor-thread-6"}
```

Does anything come to mind that is missing from the command that could relate to this?
Or potentially, could there be an issue with the kernel profile?

@gkwan-ibm

**Comment by dmikusa-pivotal** [comment]
@uhryniuk - Try using the exact command from my previous note. https://github.com/paketo-buildpacks/liberty/issues/183#issuecomment-1179346434

The kernel profile comes with basically nothing, so you have to add all the features that your app needs. Your command seems to be missing jsp-2.3. When I was playing with this before, if you were missing something the server wouldn't load the app and you'd get a 404 so I suspect that's what is happening to you.

**Comment by gkwan-ibm** [comment]
the problem is captured at issue #201

**Comment by gkwan-ibm** [comment]
also jsp is not required by the app, why need to install jsp?

**Comment by dmikusa-pivotal** [comment]
OK, so this is pending on the work from #200 also then.

>also jsp is not required by the app, why need to install jsp?

I don't honestly know, but that's the combo that worked for me, so seems like a good baseline from which to start. If it works for him also, then it's easy enough for him to play around with other options/configurations. If it doesn't work, then there's something else wrong, since that's a known good configuration.
